src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -540,5 +555,22 @@ private static bool IsRequiredImplicitNumericConversion(ITypeSymbol sourceType,
                     return false;
             }
         }
+
+        private static bool IsInUncheckedContext(CastExpressionSyntax cast)
+        {
+            SyntaxNode currentNode = cast;
+
+            do
+            {
+                if (currentNode.Kind() == SyntaxKind.UncheckedExpression || currentNode.Kind() == SyntaxKind.UncheckedStatement)
+                {
+                    return true;
+                }
+
+                currentNode = currentNode.Parent;
+            } while (currentNode is ExpressionSyntax || currentNode is StatementSyntax);
```

Comment:
Yes, Tomas is write. I'll add a test for this scenario.

I'll also fix the simplifier to check for compilation wide option and add a test for the same.